File: frontend/src/api/projects.ts
import { apiClient, unwrapData } from './client';
import { APIResponse, ApiKey, Project, RotateApiKeyResponse, Service } from '@/types/api';

export interface CreateProjectPayload {
  name: string;
  environment?: string;
  description?: string;
  logo_url?: string;
}

export interface UpdateProjectPayload {
  name?: string;
  environment?: string;
  description?: string;
  logo_url?: string;
  settings?: Record<string, any>;
}

export interface CreateApiKeyPayload {
  name: string;
  environment?: string;
}

export const projectsApi = {
  async listProjects(): Promise<Project[]> {
    const res = await apiClient.get<APIResponse<Project[]>>('/projects');
    return unwrapData(res);
  },

  async createProject(payload: CreateProjectPayload): Promise<Project> {
    const res = await apiClient.post<APIResponse<Project>>('/projects', payload);
    return unwrapData(res);
  },

  async updateProject(projectId: string, payload: UpdateProjectPayload): Promise<Project> {
    const res = await apiClient.put<APIResponse<Project>>(`/projects/${projectId}`, payload);
    return unwrapData(res);
  },

  async deleteProject(projectId: string): Promise<{ deleted: boolean }> {
    const res = await apiClient.delete<APIResponse<{ deleted: boolean }>>(`/projects/${projectId}`);
    return unwrapData(res);
  },

  async archiveProject(projectId: string): Promise<Project> {
    const res = await apiClient.post<APIResponse<Project>>(`/projects/${projectId}/archive`);
    return unwrapData(res);
  },

  async restoreProject(projectId: string): Promise<Project> {
    const res = await apiClient.post<APIResponse<Project>>(`/projects/${projectId}/restore`);
    return unwrapData(res);
  },

  async listApiKeys(projectId: string): Promise<ApiKey[]> {
    const res = await apiClient.get<APIResponse<ApiKey[]>>(`/projects/${projectId}/api-keys`);
    return unwrapData(res);
  },

  async createApiKey(projectId: string, payload: CreateApiKeyPayload): Promise<ApiKey> {
    const res = await apiClient.post<APIResponse<ApiKey>>(`/projects/${projectId}/api-keys`, payload);
    return unwrapData(res);
  },

  async revealApiKey(projectId: string, keyId: string): Promise<{ id: string; raw_key: string; revealed_at: string }> {
    const res = await apiClient.get<APIResponse<{ id: string; raw_key: string; revealed_at: string }>>(
      `/projects/${projectId}/api-keys/${keyId}/reveal`
    );
    return unwrapData(res);
  },




  async rotateApiKey(projectId: string, keyId: string): Promise<RotateApiKeyResponse> {
    const res = await apiClient.post<APIResponse<RotateApiKeyResponse>>(`/projects/${projectId}/api-keys/${keyId}/rotate`);
    return unwrapData(res);
  },

  async disableApiKey(projectId: string, keyId: string): Promise<{ disabled: boolean }> {
    const res = await apiClient.post<APIResponse<{ disabled: boolean }>>(`/projects/${projectId}/api-keys/${keyId}/disable`);
    return unwrapData(res);
  },

  async deleteApiKey(projectId: string, keyId: string): Promise<{ deleted: boolean }> {
    const res = await apiClient.delete<APIResponse<{ deleted: boolean }>>(`/projects/${projectId}/api-keys/${keyId}`);
    return unwrapData(res);
  },

  async listServices(projectId: string): Promise<Service[]> {
    const res = await apiClient.get<APIResponse<Service[]>>(`/projects/${projectId}/services`);
    return unwrapData(res);
  },
};

File: backend/api/v1/projects.py
import uuid
from typing import List
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession
from backend.api.dependencies import get_current_user, require_owner
from backend.database.session import get_async_session
from backend.models.users import User
from backend.schemas.auth import ApiKeyCreate, ApiKeyResponse, ApiKeyRevealResponse

from backend.schemas.common import APIResponse
from backend.schemas.projects import ProjectCreate, ProjectResponse, RotateApiKeyResponse, ServiceResponse, UpdateProjectRequest
from backend.services.project_service import project_service

router = APIRouter(prefix="/projects", tags=["Projects & Services"])


@router.post("", response_model=APIResponse[ProjectResponse], status_code=status.HTTP_201_CREATED)
async def create_project(
    data: ProjectCreate,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Creates a new project within the user's organization (Owner only)."""
    project = await project_service.create_project(session, current_owner.organization_id, data)
    return APIResponse(
        message="Project created successfully.",
        data=ProjectResponse.model_validate(project)
    )


@router.get("", response_model=APIResponse[List[ProjectResponse]])
async def list_projects(
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Lists projects available to the user (All projects for Owner, assigned projects for Member)."""
    projects = await project_service.list_projects(
        session,
        organization_id=current_user.organization_id,
        user_id=current_user.id,
        role=current_user.role
    )
    return APIResponse(
        message="Projects list retrieved.",
        data=[ProjectResponse.model_validate(p) for p in projects]
    )


@router.get("/{project_id}/api-keys", response_model=APIResponse[List[ApiKeyResponse]])
async def list_api_keys(
    project_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Lists SDK API keys for a project."""
    api_keys = await project_service.list_api_keys(session, project_id, current_user.organization_id)
    return APIResponse(
        message="API keys list retrieved.",
        data=[ApiKeyResponse.model_validate(k) for k in api_keys]
    )


@router.post("/{project_id}/api-keys", response_model=APIResponse[ApiKeyResponse], status_code=status.HTTP_201_CREATED)
async def create_api_key(
    project_id: uuid.UUID,
    data: ApiKeyCreate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Generates a new SDK API key for high-throughput ingestion."""
    api_key, raw_key = await project_service.create_api_key(session, project_id, current_user.organization_id, data)
    response_data = ApiKeyResponse.model_validate(api_key)
    response_data.raw_key = raw_key
    return APIResponse(
        message="SDK API Key generated. Copy the raw key now; it will not be displayed again.",
        data=response_data
    )



@router.get("/{project_id}/api-keys/{key_id}/reveal", response_model=APIResponse[ApiKeyRevealResponse])
async def reveal_api_key(
    project_id: uuid.UUID,
    key_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Reveals the decrypted raw SDK API key for authorized users."""
    from datetime import datetime, timezone
    key, raw_key = await project_service.reveal_api_key(
        session, key_id=key_id, project_id=project_id, organization_id=current_user.organization_id, user=current_user
    )
    return APIResponse(
        message="SDK API key revealed successfully.",
        data=ApiKeyRevealResponse(
            id=key.id,
            raw_key=raw_key,
            revealed_at=datetime.now(timezone.utc)
        )
    )




@router.get("/{project_id}/services", response_model=APIResponse[List[ServiceResponse]])
async def list_services(
    project_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Lists all microservices auto-discovered under a project."""
    services = await project_service.list_services(session, project_id, current_user.organization_id)
    return APIResponse(
        message="Services list retrieved.",
        data=[ServiceResponse.model_validate(s) for s in services]
    )


@router.put("/{project_id}", response_model=APIResponse[ProjectResponse])
async def update_project(
    project_id: uuid.UUID,
    data: UpdateProjectRequest,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Updates project details, logo, and settings (Owner only)."""
    project = await project_service.update_project(session, project_id, current_owner.organization_id, data)
    return APIResponse(
        message="Project updated successfully.",
        data=ProjectResponse.model_validate(project)
    )


@router.delete("/{project_id}", response_model=APIResponse[dict])
async def delete_project(
    project_id: uuid.UUID,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Soft deletes a project (Owner only)."""
    await project_service.delete_project(session, project_id, current_owner.organization_id)
    return APIResponse(
        message="Project deleted successfully.",
        data={"deleted": True}
    )


@router.post("/{project_id}/archive", response_model=APIResponse[ProjectResponse])
async def archive_project(
    project_id: uuid.UUID,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Archives a project (Owner only)."""
    project = await project_service.archive_project(session, project_id, current_owner.organization_id)
    return APIResponse(
        message="Project archived successfully.",
        data=ProjectResponse.model_validate(project)
    )


@router.post("/{project_id}/restore", response_model=APIResponse[ProjectResponse])
async def restore_project(
    project_id: uuid.UUID,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Restores an archived project (Owner only)."""
    project = await project_service.restore_project(session, project_id, current_owner.organization_id)
    return APIResponse(
        message="Project restored successfully.",
        data=ProjectResponse.model_validate(project)
    )


@router.post("/{project_id}/api-keys/{key_id}/rotate", response_model=APIResponse[RotateApiKeyResponse])
async def rotate_api_key(
    project_id: uuid.UUID,
    key_id: uuid.UUID,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Rotates an API key and returns a new raw key (Owner only)."""
    from datetime import datetime, timezone
    key, raw_key = await project_service.rotate_api_key(session, key_id, current_owner.organization_id)
    return APIResponse(
        message="API key rotated successfully. Store the new key safely.",
        data=RotateApiKeyResponse(
            id=key.id,
            name=key.name,
            prefix=key.prefix,
            environment=key.environment,
            new_raw_key=raw_key,
            rotated_at=datetime.now(timezone.utc)
        )
    )


@router.post("/{project_id}/api-keys/{key_id}/disable", response_model=APIResponse[dict])
async def disable_api_key(
    project_id: uuid.UUID,
    key_id: uuid.UUID,
    current_owner: User = Depends(require_owner),
    session: AsyncSession = Depends(get_async_session)
):
    """Disables an active API key (Owner only)."""
    await project_service.disable_api_key(session, key_id, current_owner.organization_id)
    return APIResponse(
        message="API key disabled successfully.",
        data={"disabled": True}
    )


@router.delete("/{project_id}/api-keys/{key_id}", response_model=APIResponse[dict])
async def delete_api_key(
    project_id: uuid.UUID,
    key_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session)
):
    """Deletes/revokes an active SDK API key."""
    await project_service.delete_api_key(
        session, key_id=key_id, project_id=project_id, organization_id=current_user.organization_id
    )
    return APIResponse(
        message="SDK API key revoked successfully.",
        data={"deleted": True}
    )

